README.md
# icinga-slack-bot
use slack to manage alarms in Icinga2

It can be used to interact with Icinga2 from your Slack client. It uses the
Icinga2 API to get Host/Service status details. Simple status filters can be
used to narrow down the returned status list.


## Requirements
* python >= 3.6
* python-slackclient >= 2.5.0
* certifi >= 2018
* icinga2apic >= 0.7.2
* ctparse >= 0.0.38
* Icinga2 instance with API feature enabled

## Installation

### Setting up icinga-slack-bot
* on RedHat/CentOS you need to install python3.6 and virtualenv from EPEL first
```
yum install python36-virtualenv
```

* setting up the virtual env would be done like this
```virtualenv-3.6 .pyenv``` instead of ```virtualenv .pyenv```

* here we assume we install the bot in ```/opt```
```
cd /opt
git clone https://github.com/bb-Ricardo/icinga-slack-bot.git
cd icinga-slack-bot
virtualenv .pyenv
. .pyenv/bin/activate
pip install -r requirements.txt
```

Now you would be able to start the bot with
```.pyenv/bin/python3 icinga-bot.py```

Most likely the start will fail as the config is not fully set up.
>**It is recommended to create your own config**
>```cp icinga-bot.ini.sample icinga-bot.ini```

Change config options according your environment.
After you entered the Slack tokens you should be able to start the bot.

### Run as a service
#### systemd
* a [systemd unit file](icinga-slack-bot.service) is included
but needs to be changed if the installation path is different

```
sudo cp icinga-slack-bot.service /etc/systemd/system
sudo systemctl daemon-reload
sudo systemctl start icinga-slack-bot
sudo systemctl enable icinga-slack-bot
```

#### openrc
* a [openrc init file](icinga-slack-bot.openrc) is included
but needs to be changed if the installation path is different

```
sudo install -m 755 -g root -o root icinga-slack-bot.openrc /etc/init.d/icinga-slack-bot
sudo rc-service icinga-slack-bot start
sudo rc-update add icinga-slack-bot default
```

### Run with Docker
```
git clone https://github.com/bb-Ricardo/icinga-slack-bot.git
cd icinga-slack-bot
docker build -t icinga-bot .
```

Copy the config from the [example](icinga-bot.ini.sample) to ```icinga-bot.ini``` and edit
the settings.

Now you should be able to run the image with following command
```
docker run -d -v /PATH/TO/icinga-bot.ini:/app/icinga-bot.ini --name bot icinga-bot
```

### Icinga API permissions
* necessary API permissions
  * objects/query/*
  * objects/modify/*
  * status/query
  * actions/*

This would be an Icinga Slack bot API user
```
# vim /etc/icinga2/conf.d/api-users.conf

object ApiUser "icinga-bot" {
  password = "icinga"

  permissions = [ "objects/query/*", "objects/modify/*", "status/query", "actions/*" ]
}
```

For further details check the [Icinga2 API documentation](https://icinga.com/docs/icinga2/latest/doc/12-icinga2-api)

### Get Slack Bot Token
You need to install a [Classic Slack Bot](https://api.slack.com/apps?new_classic_app=1) app
in order to use RTM.
1. Just give it a name and select the workspace.
2. Go to "App Home" and use the **Add legacy Bot User** button
3. Add a display name and the bot username like "icinga2"
4. Go to "OAuth & Permissions" and use the **Install App to Workspace** button
5. Confirm this in the bottom by selecting **Allow**
6. Use the displayed **Bot User OAuth Access Token** to add to your config

You can also use [this](contrib/icinga2_logo.png) icon to represent the bot in Slack properly.

## Configuration
icinga-slack-bot comes with a default [config file](icinga-bot.ini.sample)

## Run the bot
```
usage: icinga-bot.py [-h] [-c icinga-bot.ini] [-l {DEBUG,INFO,WARNING,ERROR}]
                     [-d]

This is an Icinga2 Slack bot.

It can be used to interact with Icinga2 from your Slack client. It uses the
Icinga2 API to get Host/Service status details. Simple status filters can be
used to narrow down the returned status list.

Version: 1.0.0 (2020-05-17)

optional arguments:
  -h, --help            show this help message and exit
  -c icinga-bot.ini, --config icinga-bot.ini
                        points to the config file to read config data from
                        which is not installed under the default path
                        './icinga-bot.ini'
  -l {DEBUG,INFO,WARNING,ERROR}, --log_level {DEBUG,INFO,WARNING,ERROR}
                        set log level (overrides config)
  -d, --daemon          define if the script is run as a systemd daemon
```

## Use the bot
Following commands are currently implemented:
* help
>display the bot help
* ping
>answers simply with pong if slack bot is running
* host status (hs)
>request a host status (or short "hs") of any or all hosts
* service status (ss)
>request a service status (or short "ss") of any or all services
* status overview (so)
>display a summary of current host and service status numbers
* acknowledge (ack)
>acknowledge problematic hosts or services
* downtime (dt)
>set a downtime for hosts/services
* comment (com)
>add a comment to hosts/services
* reschedule (rs)
>reschedule a host/service check
* send notification (sn)
>send a costum host/service notification
* delay notification (dn)
>delay host/service notifications
* show (sh)
>show a comment/downtime/acknowledgement
* remove (rm)
>remove a comment/downtime/acknowledgement
* reset (abort)
>abort current action (ack/dt/ena/disa/sh/rm)
* icinga status (is)
>print current Icinga status details
* enable (ena)
>enable an action
* disable (disa)
>disable an action

### Help command
Each command also provides a detailed help `help <command>`

### Command status filter
Following command filters are implemented
* host status (hs)
  * up
  * down
  * unreachable (unreach)
  * all
  * problems
* service status (ss)
  * ok
  * warning (warn)
  * critical (crit)
  * unknown
  * all
  * problems

Command filter can be combined like "warn crit" which would return all services in WARNING and CRITICAL state.
Also "problems" could be used to query all service with a certain name match in **NOT** OK sate.

***Important:***
* The default host status filter will only display hosts which are **NOT** UP
* The default service status filter will only display services which are **NOT** OK
* To display all host/service status just add the keyword **all** to your command

### Command name filter
You can add host names or services names to any status command.
Also just parts of host and service names can be used to search for objects

***Important:***
* when using the *service status* command only the **first two** names will be used as filter and all others are going to be ignored
* a command like `ss crit test web` will be converted into a filter like:
`
(service.state == 2) && ( match("*test*", host.name) && match("*web*", service.name) ) || ( match("*web*", host.name) && match("*test*", service.name) )
`

**HINT:**
You can also use quotes in your filter to find services containing spaces like:<br>
`ss server "network interfaces"`

#### Using previous filter
Sometime you try to tackle a problem and get tired of entering the filter every time. In this case you can use `!!` which defaults
to last commands filter.

Example:

You are checking the service status with `ss myserver ntp`. And then you want to reschedule this service.
Instead of adding the filter again you can write `rs !!` which translates into `rs myserver ntp`.
This works for all commands using filters.

### Actions
Actions have been added to perform certain actions on hosts or services.
Current actions are `Acknowledgements` and `Downtimes`.

#### Acknowledgements
This command will start a dialog to set an acknowledgement for an unhandled service or host.
This can be started with this command and the bot will ask questions about the details on following order:
1. host/service filter
2. time when acknowledgement should expire (or never)
3. a comment which should be added to the acknowledgement

##### INFO: time can be submitted in a relative format like:
_tomorrow 3pm_, _friday noon_ or _monday morning_<br>
Or more specific like _january 2nd_ or even more specific like _29.02.2020 13:00_.
Just try and see what works best for you.<br>
At the end the bot will ask you for a confirmation which can be answered with `yes` or just `y` or `no`.
After that the bot will report if the action was successful or not.

##### SORT CUT:
It's also possible to short cut the whole Q/A and just issue the action in one command:
```
ack myserver ntp until tomorrow evening Wrong ntp config, needs update
```
This will acknowledge a problematic service ntp on myserver until 6pm the following day.

##### STRUCTURE:
```
ack <host> <service> until <time> <comment>
```
or
```
ack <host> until <time> <comment>
```
or
```
ack <service> until <time> <comment>
```

#### Downtimes
This command works pretty similar to the acknowledgement command except that the bot
will ask for a downtime start. Here it's also possible to use a relative time format.

##### SORT CUT:
It's also possible to short cut the whole Q/A and just issue the action in one command:
```
dt myserver ntp from now until tomorrow evening NTP update
```
This will set a downtime for the service ntp on myserver until 6pm the following day.

##### STRUCTURE:
```
dt <host> <service> from <time> until <time> <comment>
```
or
```
dt <host> from <time> until <time> <comment>
```
or
```
dt <service> from <time> until <time> <comment>
```

### Command examples
* ```hs down test``` will display all hosts in DOWN state which match "test" as host name like "testserver" or "devtest"
* ```hs all``` will return all hosts and their status
* ```hs``` will display all hosts which currently have a problem


* ```ss warn crit ntp``` will display all services which match "ntp" and are in state CRITICAL or WARNING
* ```ss``` will display all services which currently have a problem

* ```ack myserver ntp until tomorrow evening Wrong ntp config, will be updated tomorrow``` will acknowledge a problematic service ntp on myserver until 6pm the following day

***Important:***
* The [detailed](#all-problematic-services) view will only be used if there are **1 to 4** status results

#### Help
![help example](docs/bot_help_answer.png)

#### Help for command enable
![help enable](docs/bot_help_enable.png)

#### Detailed host status example
![detailed host answer](docs/bot_detailed_host_answer.png)

#### Detailed service status example
![detailed service answer](docs/bot_detailed_service_answer.png)

#### Service name filter examples
![detailed service answer](docs/bot_host_services_status.png)
![detailed service answer](docs/bot_service_command_example.png)

#### All problematic services
![detailed service answer](docs/bot_service_status.png)

#### Status overview
![status overview answer](docs/bot_status_overview.png)

#### Acknowledge service problem
![ack_service_problem](docs/bot_ack_service.png)

#### Disable notifications global
![disable_notifications_global](docs/bot_disable_notifications_global.png)

#### Icinga status
![icinga_status](docs/bot_icinga_status.png)

### Startup messages
* once the bot starts it will report a short status to the configured default channel
![bot started successfully](docs/bot_start_success.png)
![bot had issues starting](docs/bot_start_failure.png)

## Alert notification
To get Slack notifications if something goes wrong you can check out the
notification handlers in [contrib](contrib).

Copy [slack-notification.sh](contrib/slack-notification.sh) to `/etc/icinga2/scripts/`
and add the [icinga2_slack_notification_commands.conf](contrib/icinga2_slack_notification_commands.conf)
to your icinga configuration (depends on your where you need to put this)

More information here: [Icinga2 Docs -> notifications](https://icinga.com/docs/icinga2/latest/doc/03-monitoring-basics/#notifications).

Don't forget to set `NOTIFICATION_CONFIG` in
*icinga2_slack_notification_commands.conf* to the full path to your bot config file.

### Slack WEBHOOK URL
In order to send alarms to Slack you need a Webhook URL.
1. In your Slack Bot settings use **Incoming Webhooks** under **Features**.
2. Switch **Activate Incoming Webhooks** to **On**
3. Use **Add New Webhook to Workspace**
4. Select the channel the alert messages should be posted to.
5. Now you should have a Webhook URL you need to copy to your configuration


### Alert examples
![alert host down](docs/notification_host_down.png)
![alert host up](docs/notification_host_up.png)
![alert service problem](docs/notification_service_problem.png)


## License
>You can check out the full license [here](LICENSE.txt)

This project is licensed under the terms of the **MIT** license.

### Saying Thank You
quite some inspiration came from [mlabouardy](https://github.com/mlabouardy) and his Go implementation
of a [slack bot](https://github.com/mlabouardy/icinga2-slack-bot)


CHANGELOG.md
# Changelog

[1.0.0](https://github.com/bb-Ricardo/icinga-slack-bot/tree/v1.0.0) (2020-05-17) *The Baby learned to walk*

I took quite some time but here it is: The first major release 🎉

This release adds most of the Icinga2 API actions which can be performed on objects.
Also handling comments, downtimes and acknowledgements are now fully supported.

**Features:**
* Reschedule host or service [#26](https://github.com/bb-Ricardo/icinga-slack-bot/issues/26)
* Implement enable/disable command to modify application features [#24](https://github.com/bb-Ricardo/icinga-slack-bot/issues/24)
* Add/display comments, acknowledgements and downtimes to host and service status results [#25](https://github.com/bb-Ricardo/icinga-slack-bot/issues/25)
* Add ability to remove acknowledgements/downtimes/comments [#29](https://github.com/bb-Ricardo/icinga-slack-bot/issues/29)
* Add command to display services/hosts which are ack/dt/com [#38](https://github.com/bb-Ricardo/icinga-slack-bot/issues/38)
* Allow to use previous command options with special option „!!“ [#37](https://github.com/bb-Ricardo/icinga-slack-bot/issues/37)
* Add action to delay notifications [#33](https://github.com/bb-Ricardo/icinga-slack-bot/issues/33)
* Add comments and downtime details to object details view [#31](https://github.com/bb-Ricardo/icinga-slack-bot/issues/31)
* Add “💬” to host and service title if these objects have comments [#30](https://github.com/bb-Ricardo/icinga-slack-bot/issues/30)
* Add „(handled)“ to host/service links [#28](https://github.com/bb-Ricardo/icinga-slack-bot/issues/28)
* Add command to send custom notification [#27](https://github.com/bb-Ricardo/icinga-slack-bot/issues/27)
* Query user details delayed [#23](https://github.com/bb-Ricardo/icinga-slack-bot/issues/23)
* Parse filters with quotes differently to allow for explicit names [#18](https://github.com/bb-Ricardo/icinga-slack-bot/issues/18)
* Tell user on error which command the user is using [#40](https://github.com/bb-Ricardo/icinga-slack-bot/issues/40)

**Bug fixes:**
* Take care of pending services without check result [#39](https://github.com/bb-Ricardo/icinga-slack-bot/issues/39)

**Documentation**
* Add description of slack bot creation to README [#35](https://github.com/bb-Ricardo/icinga-slack-bot/issues/35)

**Internal:**
* Refactor internal command parsing and create class with command methods [#19](https://github.com/bb-Ricardo/icinga-slack-bot/issues/19)

**Dependencies**
* Move to icinga2apic as dependency [#32](https://github.com/bb-Ricardo/icinga-slack-bot/issues/32)


[0.2.0](https://github.com/bb-Ricardo/icinga-slack-bot/tree/0.2.0) (2019-11-16) *Actions Arrived*

**Features:**
* Added action to acknowledge problematic hosts/services [#1](https://github.com/bb-Ricardo/icinga-slack-bot/issues/1)
* Added action to set a downtime for hosts/services [#4](https://github.com/bb-Ricardo/icinga-slack-bot/issues/4)
* Added a status overview command [#2](https://github.com/bb-Ricardo/icinga-slack-bot/issues/2)
* Added Dockerfile [#3](https://github.com/bb-Ricardo/icinga-slack-bot/issues/3)
* Added Icinga status command [#5](https://github.com/bb-Ricardo/icinga-slack-bot/issues/5)
* Added config option to limit max returned results [#7](https://github.com/bb-Ricardo/icinga-slack-bot/issues/7)
* Handled hosts/services are not displayed any longer with default host/service status command [#6](https://github.com/bb-Ricardo/icinga-slack-bot/issues/6)
* Added filter option to config file to limit results to single hosts or host groups [#8](https://github.com/bb-Ricardo/icinga-slack-bot/issues/8)
* Added a more detailed help, help can be called for each command [#9](https://github.com/bb-Ricardo/icinga-slack-bot/issues/9)

**Bug fixes:**
* host state not reporting hosts in NOT UP state while running default query [#15](https://github.com/bb-Ricardo/icinga-slack-bot/issues/15)

**Internal:**
* Splitted code into modules [#12](https://github.com/bb-Ricardo/icinga-slack-bot/issues/12)


[0.1.0](https://github.com/bb-Ricardo/icinga-slack-bot/tree/0.1.0) (2019-07-03) *Initial Release*

**Road to Initial Release**
* check icinga connection
* configure default channel to post after bot start
* proper message formatting
* add systemd unit file
* add README.md
* fix logging
* document functions
* add debug logging
* add description
* proper icinga2 error handling
* add all icinga2api connection options (ca, cert, key, timeout)
* add icingaweb2 url to bot responses and error handler
* circumvent *"[ERROR] block must be less than 3000 characters"*
* add state change time to host and service status messages (detailed view)
* add fallback text to slack messages to display push messages
* use attachments to display bot commands


i2_slack_modules/common.py
####
#
# Some commonly used functions
#

import configparser
import logging
from argparse import ArgumentParser, RawDescriptionHelpFormatter
from datetime import datetime
import inspect
import re

from ctparse import ctparse


# define valid log levels
valid_log_levels = ["DEBUG", "INFO", "WARNING", "ERROR"]


def parse_command_line(version=None,
                       self_description=None,
                       version_date=None,
                       default_config_file_path=None,
                       ):
    """parse command line arguments

    Also add current version and version date to description
    """

    # define command line options
    description = "%s\nVersion: %s (%s)" % (self_description, version, version_date)

    parser = ArgumentParser(
        description=description,
        formatter_class=RawDescriptionHelpFormatter)

    parser.add_argument("-c", "--config", default=default_config_file_path, dest="config_file",
                        help="points to the config file to read config data from " +
                             "which is not installed under the default path '" +
                             default_config_file_path + "'",
                        metavar="icinga-bot.ini")
    parser.add_argument("-l", "--log_level", choices=valid_log_levels, dest="log_level",
                        help="set log level (overrides config)")
    parser.add_argument("-d", "--daemon",  action='store_true', dest="daemon",
                        help="define if the script is run as a systemd daemon")

    return parser.parse_args()


def parse_own_config(args, default_log_level):
    """parsing and basic validation of own config file

    Parameters
    ----------
    args : ArgumentParser object

    default_log_level: str
        default log level if log level is not set in config

    Returns
    -------
    dict
        a dictionary with all config options parsed from the config file
    """

    config_dict = {}

    config_error = False

    config_file = args.config_file

    logging.debug("Parsing daemon config file: %s" % config_file)

    if config_file is None or config_file == "":
        do_error_exit("Config file not defined.")

    if default_log_level is None or default_log_level == "":
        do_error_exit("Default log level not set.")

    # setup config parser and read config
    config_handler = configparser.ConfigParser(strict=True, allow_no_value=True)

    # noinspection PyBroadException
    try:
        config_handler.read_file(open(config_file))
    except configparser.Error as e:
        do_error_exit("Error during config file parsing: %s" % e)
    # noinspection PyBroadException
    except Exception:
        do_error_exit("Unable to open file '%s'" % config_file)

    # read logging section
    this_section = "main"
    if this_section not in config_handler.sections():
        logging.warning("Section '%s' not found in '%s'" % (this_section, config_file))

    # read logging if present
    config_dict["log_level"] = config_handler.get(this_section, "log_level", fallback=default_log_level)

    # overwrite log level with command line argument
    if args.log_level is not None and args.log_level != "":
        config_dict["log_level"] = args.log_level
        logging.info("Config: overwriting log_level with command line arg: %s" % args.log_level)

    # set log level again
    if args.log_level is not config_dict["log_level"]:
        set_log_level(config_dict["log_level"])

    logging.debug("Config: %s = %s" % ("log_level", config_dict["log_level"]))

    # read common section
    this_section = "slack"
    if this_section not in config_handler.sections():
        do_error_exit("Section '%s' not found in '%s'" % (this_section, config_file))
    else:
        config_dict["slack.bot_token"] = config_handler.get(this_section, "bot_token", fallback="")
        logging.debug("Config: %s = %s***" % ("slack.bot_token", config_dict["slack.bot_token"][0:10]))
        config_dict["slack.default_channel"] = config_handler.get(this_section, "default_channel", fallback="")
        logging.debug("Config: %s = %s" % ("slack.default_channel", config_dict["slack.default_channel"]))

    # read paths section
    this_section = "icinga"
    if this_section not in config_handler.sections():
        do_error_exit("Section '%s' not found in '%s'" % (this_section, config_file))
    else:
        config_dict["icinga.hostname"] = config_handler.get(this_section, "hostname", fallback="")
        logging.debug("Config: %s = %s" % ("icinga.hostname", config_dict["icinga.hostname"]))
        config_dict["icinga.port"] = config_handler.get(this_section, "port", fallback="")
        logging.debug("Config: %s = %s" % ("icinga.port", config_dict["icinga.port"]))
        config_dict["icinga.username"] = config_handler.get(this_section, "username", fallback="")
        logging.debug("Config: %s = %s" % ("icinga.username", config_dict["icinga.username"]))
        config_dict["icinga.password"] = config_handler.get(this_section, "password", fallback="")
        logging.debug("Config: %s = %s***" % ("icinga.password", config_dict["icinga.password"][0:3]))
        config_dict["icinga.web2_url"] = config_handler.get(this_section, "web2_url", fallback="")
        logging.debug("Config: %s = %s" % ("icinga.web2_url", config_dict["icinga.web2_url"]))
        config_dict["icinga.certificate"] = config_handler.get(this_section, "certificate", fallback="")
        logging.debug("Config: %s = %s" % ("icinga.certificate", config_dict["icinga.certificate"]))
        config_dict["icinga.key"] = config_handler.get(this_section, "key", fallback="")
        logging.debug("Config: %s = %s" % ("icinga.key", config_dict["icinga.key"]))
        config_dict["icinga.ca_certificate"] = config_handler.get(this_section, "ca_certificate", fallback="")
        logging.debug("Config: %s = %s" % ("icinga.ca_certificate", config_dict["icinga.ca_certificate"]))
        config_dict["icinga.timeout"] = config_handler.get(this_section, "timeout", fallback="")
        logging.debug("Config: %s = %s" % ("icinga.timeout", config_dict["icinga.timeout"]))
        config_dict["icinga.filter"] = config_handler.get(this_section, "filter", fallback="")
        logging.debug("Config: %s = %s" % ("icinga.filter", config_dict["icinga.filter"]))
        config_dict["icinga.max_returned_results"] = \
            config_handler.get(this_section, "max_returned_results", fallback="")
        logging.debug("Config: %s = %s" % ("icinga.max_returned_results", config_dict["icinga.max_returned_results"]))

    for key, value in config_dict.items():
        if value is "":
            # if we use a certificate then don't care if user or password are defined
            if key in ["icinga.username", "icinga.password"] and config_dict["icinga.certificate"] != "":
                continue
            # these vars can be empty
            if key in ["icinga.key", "icinga.certificate", "icinga.web2_url", "icinga.ca_certificate",
                       "icinga.filter", "icinga.max_returned_results", "icinga.timeout"]:
                continue
            logging.error("Config: option '%s' undefined or empty!" % key)
            config_error = True

    if config_error:
        return False

    return config_dict


def do_error_exit(log_text):
    """log an error and exit with return code 1

    Parameters
    ----------
    log_text : str
        the text to log as error
    """

    logging.error(log_text)
    exit(1)


def set_log_level(log_level=None):
    """set or reset the log level

    Parameters
    ----------
    log_level : str
        Log level to set

    """

    # check set log level against self defined log level array
    if not log_level.upper() in valid_log_levels:
        do_error_exit('Invalid log level: %s' % log_level)

    # check the provided log level and bail out if something is wrong
    numeric_log_level = getattr(logging, log_level.upper(), None)
    if not isinstance(numeric_log_level, int):
        do_error_exit('Invalid log level: %s' % log_level)

    logging.info("Setting log level to: %s" % log_level)

    # unfortunately we have to manipulate the root logger
    if log_level == "DEBUG":
        logging.disable(logging.NOTSET)
    elif log_level == "INFO":
        logging.disable(logging.DEBUG)
    elif log_level == "WARNING":
        logging.disable(logging.INFO)
    elif log_level == "ERROR":
        logging.disable(logging.WARNING)


def setup_logging(args=None, default_log_level=None):
    """Setup logging

    Parameters
    ----------
    args : ArgumentParser object

    default_log_level: str
        default log level if args.log_level is not set

    """

    log_level = args.log_level

    if args.daemon:
        # omit time stamp if run in daemon mode
        logging.basicConfig(level="DEBUG", format='%(levelname)s: %(message)s')
    else:
        logging.basicConfig(level="DEBUG", format='%(asctime)s - %(levelname)s: %(message)s')

    if log_level is None or log_level == "":
        logging.debug("Configuring logging: No log level defined, using default level: %s" % default_log_level)
        log_level = default_log_level

    set_log_level(log_level)


def ts_to_date(ts, date_format="%Y-%m-%d %H:%M:%S"):
    """
    Return a formatted date/time string from a given time stamp

    Parameters
    ----------
    ts : int, float
        time stamp to convert
    date_format : string
        format to convert time stamp to

    Returns
    -------
    str: formatted date/time string
    """
    return datetime.fromtimestamp(ts).strftime(date_format)


def parse_relative_date(data_to_parse=None):
    """
    Return a ctparse.Time dict and a datetime object for a string of relative date and/or time to parse.

    Parameters
    ----------
    data_to_parse : string
        string with relative time information which should be parsed into absolute datetime object

    Returns
    -------
    dict: date/time data + datetime object

        example_output = {
            'mstart': 0,
            'mend': 8,
            'year': 2019,
            'month': 11,
            'day': 5,
            'hour': 17,
            'minute': 33,
            'DOW': None,
            'POD': None,
            'dt': datetime.datetime(2019, 11, 5, 17, 33)
        }
    """

    string_to_parse = None
    if isinstance(data_to_parse, list):
        string_to_parse = " ".join(data_to_parse)

    elif isinstance(data_to_parse, str):
        string_to_parse = data_to_parse

    if string_to_parse is None:
        logging.warning("Trying to parse date but submitted data is not a string or a list.")
        return None

    logging.debug("%s START ctparse %s" % ("*" * 10, "*" * 50))
    parsed_date = ctparse(string_to_parse)
    logging.debug("%s END ctparse %s" % ("*" * 10, "*" * 52))

    if parsed_date is None or parsed_date.resolution is None:
        logging.debug("Unable to parse a date from string: %s" % string_to_parse)
        return None

    data_parts = parsed_date.resolution

    # just do some own additional parsing
    time_string = string_to_parse[data_parts.mstart:data_parts.mend]

    if any(keyword in time_string for keyword in ["lunch", "noon", "mittag"]):
        data_parts.hour = 12

    if "morning" in time_string:
        data_parts.hour = 9

    if "afternoon" in time_string:
        data_parts.hour = 15

    if "evening" in time_string:
        data_parts.hour = 18

    # unable to determine time of the day
    # use current time
    if data_parts.hour is None:
        now = datetime.today()
        data_parts.hour = now.hour
        data_parts.minute = now.minute

    # if minute returned None set to full hour
    if data_parts.minute is None:
        data_parts.minute = 0

    dt = None
    try:
        dt = datetime(year=data_parts.year,
                      month=data_parts.month,
                      day=data_parts.day,
                      hour=data_parts.hour,
                      minute=data_parts.minute)
    except TypeError:
        pass

    if data_parts:
        logging.debug("Parsed date from string (%s): %s" %
                      (string_to_parse[data_parts.mstart:data_parts.mend], parsed_date))
    else:
        logging.debug("Unable to parse a date from string: %s" % string_to_parse)

    return_data = data_parts.__dict__
    return_data["dt"] = dt

    return return_data


def my_own_function_name():
    """returns the name of the function who called this function"""
    return inspect.currentframe().f_back.f_code.co_name


def quoted_split(string_to_split, preserve_quotations=False):
    """
    Function to split string into list of words but preserves quoted
    sub strings.

    found here: https://stackoverflow.com/a/51560564

    Parameters
    ----------
    string_to_split : string
        the string to split
    preserve_quotations : bool
        determine if quote marks should be preserved or not

    Returns
    -------
    list: splitted string as list
    """

    def strip_quotes(s):
        if s and (s[0] == '"' or s[0] == "'") and s[0] == s[-1] and preserve_quotations is False:
            return s[1:-1].replace('\\"', '"').replace("\\'", "'")
        return s
    return [strip_quotes(p)
            for p in re.findall(r'"(?:\\.|[^"])*"|\'(?:\\.|[^\'])*\'|[^\s]+', string_to_split)]
